lib/evaluate/rte.cc

```diff
@@ -0,0 +1,166 @@
+// real arg for n
+template<typename HostT> static HostT Bessel_jn(std::int64_t n, HostT x) {
+  return std::cyl_bessel_j(static_cast<HostT>(n), x);
+}
+template<typename HostT> static HostT Bessel_yn(std::int64_t n, HostT x) {
+  return std::cyl_neumann(static_cast<HostT>(n), x);
+}
+
+template<typename HostT> void AddLibmRealHostProcedure(HostRte &hostRte) {
+  using F = FuncPointer<HostT, HostT>;
+  using F2 = FuncPointer<HostT, HostT, HostT>;
+  HostRteProcedureSymbol libmSymbols[]{{"acos", F{std::acos}, true},
+      {"acosh", F{std::acosh}, true}, {"asin", F{std::asin}, true},
+      {"asinh", F{std::asinh}, true}, {"atan", F{std::atan}, true},
+      {"atan", F2{std::atan2}, true}, {"atanh", F{std::atanh}, true},
+      {"bessel_jn", &Bessel_jn<HostT>, true},
+      {"bessel_yn", &Bessel_yn<HostT>, true}, {"cos", F{std::cos}, true},
+      {"cosh", F{std::cosh}, true}, {"sin", F{std::sin}, true},
+      {"sinh", F{std::sinh}, true}, {"sqrt", F{std::sqrt}, true},
+      {"tan", F{std::tan}, true}, {"tanh", F{std::tanh}, true}};
+
+  for (auto sym : libmSymbols) {
+    hostRte.AddProcedure(std::move(sym));
+  }
+}
+
+template<typename HostT> void AddLibmComplexHostProcedure(HostRte &hostRte) {
+  using F = FuncPointer<std::complex<HostT>, const std::complex<HostT> &>;
+  HostRteProcedureSymbol libmSymbols[]{{"acos", F{std::acos}, true},
+      {"acosh", F{std::acosh}, true}, {"asin", F{std::asin}, true},
+      {"asinh", F{std::asinh}, true}, {"atan", F{std::atan}, true},
+      {"atanh", F{std::atanh}, true}, {"cos", F{std::cos}, true},
+      {"cosh", F{std::cosh}, true}, {"sin", F{std::sin}, true},
+      {"sinh", F{std::sinh}, true}, {"sqrt", F{std::sqrt}, true},
+      {"tan", F{std::tan}, true}, {"tanh", F{std::tanh}, true}};
+
+  for (auto sym : libmSymbols) {
+    hostRte.AddProcedure(std::move(sym));
+  }
+}
+
+void HostRte::DefaultInit() {
+  AddLibmRealHostProcedure<float>(*this);
+  AddLibmRealHostProcedure<double>(*this);
+  AddLibmRealHostProcedure<long double>(*this);
+
+  AddLibmComplexHostProcedure<float>(*this);
+  AddLibmComplexHostProcedure<double>(*this);
+  AddLibmComplexHostProcedure<long double>(*this);
+}
+
+void HostFloatingPointEnvironment::SetUpHostFloatingPointEnvironment(
+    FoldingContext &context) {
+  errno = 0;
+  if (feholdexcept(&originalFenv_) != 0) {
+    context.messages().Say(
+        "Folding with host runtime: feholdexcept() failed: %s"_en_US,
+        std::strerror(errno));
+    return;
+  }
+  if (fegetenv(&currentFenv_) != 0) {
+    context.messages().Say(
+        "Folding with host runtime: fegetenv() failed: %s"_en_US,
+        std::strerror(errno));
+    return;
+  }
+#if __x86_64__
+  if (context.flushSubnormalsToZero()) {
+    currentFenv_.__mxcsr |= 0x8000;  // result
+    currentFenv_.__mxcsr |= 0x0040;  // operands
+  } else {
+    currentFenv_.__mxcsr &= ~0x8000;  // result
+    currentFenv_.__mxcsr &= ~0x0040;  // operands
+  }
+#else
+  // TODO other architecture
```

Comment:
Maybe we should have a #warning here.

Reply:
I went for a runtime TODO warning in commit f776c8d. I do not want to annoy people for a feature they may not use.

Reply:
For AArch64 architecture, please either look into merged flang's pull requests https://github.com/flang-compiler/flang/pull/488 and https://github.com/flang-compiler/flang/pull/489 and also into two of my comments under f18's bug report https://github.com/flang-compiler/f18/issues/148 (https://github.com/flang-compiler/f18/issues/148#issuecomment-408269774 and https://github.com/flang-compiler/f18/issues/148#issuecomment-409990207) or discuss and synchronize addition of missing implementation during our regular calls.